CASSANDRA-17334 Pre hashed passwords in CQL for trunk

bereng opened the pull request.

bereng committed:
Pre hashed passwords in CQL

bereng committed:
CIRCLE TO MID DO NOT MERGE

bereng committed:
CIRCLE TO 15234 CCM PATCHED DO NOT MERGE

bereng commented:
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/605/workflows/5ec22385-2fb8-478b-993c-fc0739a5151d)
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/605/workflows/4b5c9990-ced5-4b5b-a656-e8f1b2203706)

adelapena commented on src/java/org/apache/cassandra/auth/CassandraRoleManager.java:
```diff
@@ -172,14 +172,14 @@ protected final void loadRoleStatement()
                                          options.getSuperuser().or(false),
                                          options.getLogin().or(false),
-                                         escape(hashpw(options.getPassword().get())))
+                                         options.getHashedPassword().or(() -> escape(hashpw(options.getPassword().get()))))
```
Guava's `Optional#or` has been marked as beta for a while. I think that `RoleOptions` uses Guava's `Optional` because it was created before Java 8, now we can easily replace it with `java.util.Optional`. I gave it a try [here](https://github.com/adelapena/cassandra/commit/f6bf02fd02aa58ae99f15b4865a212204d2ee907) and the change seems pretty straightforward. I know this is not strictly related to the patch but since we are here...

bereng replied:
sgtm

adelapena commented on src/java/org/apache/cassandra/auth/RoleOptions.java:
```diff
@@ -89,6 +90,15 @@ public boolean isEmpty()
+     */
+    public Optional<String> getHashedPassword()
+    {
+        return Optional.fromNullable((String)options.get(IRoleManager.Option.HASHED_PASSWORD));
```
Nit: missed whitespace
```suggestion
        return Optional.fromNullable((String) options.get(IRoleManager.Option.HASHED_PASSWORD));
```

adelapena commented on src/java/org/apache/cassandra/cql3/PasswordObfuscator.java:
```diff
@@ -62,10 +62,15 @@ public static String obfuscate(String query, RoleOptions opts)
+        //  - Match new line and case insensitive (?si), and PASSWORD_TOKEN with greedy mode up to the start of the actual password and group it.
+        //  - Quote the password between \Q and \E so any potential special characters are ignored
+        //  - Replace the match with the grouped data + the obfuscated token
+        return query.replaceAll("((?si)"+ PASSWORD_TOKEN + ".+?)\\Q" + pass.get() + "\\E", "$1" + PasswordObfuscator.OBFUSCATION_TOKEN);
```
Nit: I would split the arguments for better readability, feel free to ignore if you don't agree:
```suggestion
        return query.replaceAll("((?si)"+ PASSWORD_TOKEN + ".+?)\\Q" + pass.get() + "\\E",
                                "$1" + PasswordObfuscator.OBFUSCATION_TOKEN);
```

adelapena commented on src/java/org/apache/cassandra/tools/HashPassword.java:
```diff
@@ -0,0 +1,195 @@
+                    }
+                    catch (IOException e)
+                    {
+                        System.err.println(String.format("Failed to read from '%s': %s",
```
No need to break the line. Also, we could use `System.err.printf("Failed to read from '%s': %s%n", input, e);`.

adelapena commented on src/java/org/apache/cassandra/tools/HashPassword.java:
```diff
@@ -0,0 +1,195 @@
+            if (password.chars().anyMatch(i -> i < 32))
+                System.err.println("WARNING: The provided plain text password contains non-printable characters (ASCII<32).");
+
+            if (password.length() < 4)
```
I would place that `4` in a class constant. We don't check password lengths when creating a user/role with a plain password, do we?

smiklosovic replied:
Why do we even check the length and why it has to be 4 and not 3? Or any number for that matter. What is this "4" based on? Good faith / practice? I do not think doing this check in this PR is a right thing to do. It should be probably done in a separate ticket with more discussion what we want to actually achieve.

I can imagine that it would need to pass some filter, like it has to contain number, upper and lowercase characters and special symbol and so on ... There might be some policy around what passwords are acceptable and it would refuse to hash it (or it would refuse to create role with such password) when created in CQLSH.

Hence, in this PR, I would not impose any restrictions on how the password might look like untile we agree on something more robust.

bereng replied:
We don't actually. And it would be a breaking change only fit for a major I believe but worth of another ticket imo. wdyt?

smiklosovic replied:
Yeah I agree.

bereng replied:
Oh sorry, our replies crossed. It is not a restriction atm, just a friendly warning. I mean it prints the message but will still hash it for you. I would leave it as it doesn't hurt and it's a good reminder. Then yes a new ticket as we said to address all those issues.

```
tools/bin/hash_password --plain A
WARNING: The provided password is very short, probably too short to be secure.
$2a$10$2hfvBzy2skr0Kpfp0Q5rXOGGVjt4fhY2hp2vyF1aHttzDrSkb0/Hq
```

adelapena commented on test/unit/org/apache/cassandra/tools/ToolRunner.java:
```diff
@@ -136,6 +138,8 @@ public void run()
                     int read = input.read(buffer);
                     if (read == -1)
                     {
+                        if(autoCloseOut)
```
Nit: missed whitespace
```suggestion
                        if (autoCloseOut)
```

adelapena commented on test/unit/org/apache/cassandra/tools/ToolRunner.java:
```diff
@@ -119,11 +119,13 @@ public void checkPermission(Permission perm, Object context)
 
         private final InputStream input;
         private final T out;
+        private boolean autoCloseOut;
```
Nit
```suggestion
        private final boolean autoCloseOut;
```

adelapena commented on test/unit/org/apache/cassandra/tools/ToolRunner.java:
```diff
@@ -119,11 +119,13 @@ public void checkPermission(Permission perm, Object context)
+        private boolean autoCloseOut;
 
-        private StreamGobbler(InputStream input, T out)
+        private StreamGobbler(InputStream input, T out, boolean autoCloseOut)
```
Is this one of the two bugs mentioned on the ticket?

bereng replied:
Yes, the process would just spin waiting for more input, it was a bug in `ToolRunner`

adelapena reviewed (commented):
Left some initial comments, I'm still reviewing but so far looks good :)

adelapena commented on test/unit/org/apache/cassandra/audit/AuditLoggerAuthTest.java:
```diff
@@ -143,6 +144,21 @@ public void testCqlCreateRoleSyntaxError()
+    public void testCqlCreateRoleSyntaxErrorWithHashedPwd()
+    {
+        String createTestRoleCQL = String.format("CREATE ROLE %s WITH LOGIN = %s ANDSUPERUSER = %s AND HASHED PASSWORD",
+                                                 TEST_ROLE, true, false) + CASS_PW;
```
Doesn't change the test, but it would be `TEST_PW_HASH` instead of `CASS_PW`.

bereng replied:
Good catch!

adelapena commented on test/unit/org/apache/cassandra/audit/AuditLoggerAuthTest.java:
```diff
@@ -143,6 +144,21 @@ public void testCqlCreateRoleSyntaxError()
+        String createTestRoleCQLExpected = String.format("CREATE ROLE %s WITH LOGIN = %s ANDSUPERUSER = %s AND HASHED PASSWORD ",
+                                                         TEST_ROLE, true, false) + PasswordObfuscator.OBFUSCATION_TOKEN;
+
+        executeWithCredentials(Arrays.asList(createTestRoleCQL), CASS_USER, CASS_PW, AuditLogEntryType.LOGIN_SUCCESS);
```
```suggestion
        executeWithCredentials(Collections.singletonList(createTestRoleCQL), CASS_USER, CASS_PW, AuditLogEntryType.LOGIN_SUCCESS);
```

adelapena commented on test/unit/org/apache/cassandra/auth/CreateAndAlterRoleTest.java:
```diff
@@ -0,0 +1,154 @@
+        assertInvalidMessage("Options 'password' and 'hashed password' are mutually exclusive",
+                             String.format("CREATE ROLE %s WITH login=true AND password='%s' AND hashed password='%s'",
+                                           user1, plainTextPwd, hashedPassword));
+        executeNet(String.format("CREATE ROLE %s WITH login=true AND hashed password='%s'",
+                                 user1, hashedPassword));
+        executeNet(String.format("CREATE ROLE %s WITH login=true AND password='%s'",
+                                 user2, plainTextPwd));
```
Nit: no need to break this lines
```suggestion
        executeNet(String.format("CREATE ROLE %s WITH login=true AND hashed password='%s'", user1, hashedPassword));
        executeNet(String.format("CREATE ROLE %s WITH login=true AND password='%s'", user2, plainTextPwd));
```

adelapena commented on test/unit/org/apache/cassandra/auth/CreateAndAlterRoleTest.java:
```diff
@@ -0,0 +1,154 @@
+        assertInvalidMessage("Options 'password' and 'hashed password' are mutually exclusive",
+                             String.format("ALTER ROLE %s WITH password='%s' AND hashed password='%s'",
+                                           user1, plainTextPwd2, hashedPassword2));
+        executeNet(String.format("ALTER ROLE %s WITH password='%s'",
+                                 user1, plainTextPwd2));
+        executeNet(String.format("ALTER ROLE %s WITH hashed password='%s'",
+                                 user2, hashedPassword2));
```
Nit: no need to break these lines
```suggestion
        executeNet(String.format("ALTER ROLE %s WITH password='%s'", user1, plainTextPwd2));
        executeNet(String.format("ALTER ROLE %s WITH hashed password='%s'", user2, hashedPassword2));
```

adelapena commented on test/unit/org/apache/cassandra/auth/CreateAndAlterRoleTest.java:
```diff
@@ -0,0 +1,154 @@
+        assertInvalidMessage("Invalid hashed password value",
+                             String.format("CREATE USER %s WITH hashed password '%s'",
+                                           user1, "this_is_an_invalid_hash"));
+        executeNet(String.format("CREATE USER %s WITH hashed password '%s'",
+                                 user1, hashedPassword));
+        executeNet(String.format("CREATE USER %s WITH password '%s'",
+                                 user2, plainTextPwd));
```
Nit: no need to break these lines
```suggestion
        executeNet(String.format("CREATE USER %s WITH hashed password '%s'", user1, hashedPassword));
        executeNet(String.format("CREATE USER %s WITH password '%s'",  user2, plainTextPwd));
```

adelapena commented on test/unit/org/apache/cassandra/auth/CreateAndAlterRoleTest.java:
```diff
@@ -0,0 +1,154 @@
+                executeNet(query);
+                return true;
+            }
+            catch(Throwable e)
```
We could verify that the thrown exception is `com.datastax.driver.core.exceptions.AuthenticationException`.

adelapena replied:
```suggestion
            catch (Throwable e)
```

adelapena commented on test/unit/org/apache/cassandra/tools/HashPasswordTest.java:
```diff
@@ -0,0 +1,133 @@
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    private String plaintextPassword = "foobar";
+    private String hashPasswordTool = "tools/bin/hash_password";
```
Could be static final

adelapena commented on test/unit/org/apache/cassandra/tools/HashPasswordTest.java:
```diff
@@ -0,0 +1,133 @@
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class HashPasswordTest extends CQLTester
```
Nice test! I think we are missing the `-r` argument and passwords with less than 4 chars.

adelapena commented on test/unit/org/apache/cassandra/tools/HashPasswordTest.java:
```diff
@@ -0,0 +1,133 @@
+    {
+        ToolResult tool = ToolRunner.invoke(Collections.emptyMap(),
+                                            new ByteArrayInputStream(plaintextPassword.getBytes()),
+                                            Arrays.asList(hashPasswordTool, "--input", "-"));
```
What's the purpose of the `"-"` argument?

bereng replied:
It's in the help `Input is a file (or - for stdin)`

adelapena commented on src/java/org/apache/cassandra/auth/RoleOptions.java:
```diff
@@ -134,6 +144,26 @@ public void validate()
+                    }
+                    catch (Exception e)
+                    {
+                        throw new InvalidRequestException("Invalid hashed password value");
```
Since we don't have an easy way to produce hashes on client libraries, maybe we could mention here what library should be used to generate the hash (jbcrypt)? Otherwise we could find frustrated users trying to use other bcrypts.

adelapena commented on test/unit/org/apache/cassandra/auth/CreateAndAlterRoleTest.java:
```diff
@@ -0,0 +1,157 @@
+
+        useSuperUser();
+
+        executeNet(String.format("ALTER USER %s WITH password '%s'",
+                                 user1, plainTextPwd2));
+        executeNet(String.format("ALTER USER %s WITH hashed password '%s'",
+                                 user2, hashedPassword2));
```
Nit: these lines don't need to be broken:
```suggestion
        executeNet(String.format("ALTER USER %s WITH password '%s'", user1, plainTextPwd2));
        executeNet(String.format("ALTER USER %s WITH hashed password '%s'", user2, hashedPassword2));
```

adelapena commented on test/unit/org/apache/cassandra/auth/CreateAndAlterRoleTest.java:
```diff
@@ -0,0 +1,157 @@
+            }
+            catch (Throwable e)
+            {
+                if (e instanceof AuthenticationException)
+                {
+                    reinitializeNetwork();
+                    return false;
+                }
+                else
+                    fail("Unexpected exception: " + e);
+            }
+
+            return false;
```
We could simplify and just use an assertion, that should be enough to stop the spinning.
```suggestion
                assertTrue("Unexpected exception: " + e, e instanceof AuthenticationException);
                reinitializeNetwork();
                return false;
            }
```

adelapena commented on test/unit/org/apache/cassandra/tools/HashPasswordTest.java:
```diff
@@ -0,0 +1,150 @@
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    private final String plaintextPassword = "foobar";
```
Could be static:
```suggestion
    private static final String plaintextPassword = "foobar";
```

adelapena commented on test/unit/org/apache/cassandra/tools/HashPasswordTest.java:
```diff
@@ -0,0 +1,150 @@
+                        "/foo/bar/baz/blah/yadda");
+    }
+
+    private void assertToolError(String expectedMessage, String... args)
```
Could be static
```suggestion
    private static void assertToolError(String expectedMessage, String... args)
```

adelapena reviewed (approved):
Looks good to me, just left a couple of additional very minor suggestions that can be addressed on commit, if you agree.

bereng committed:
CIRCLE BACK TO STD MID

bereng commented:
Rebased and fixed a failing test pre-commit:
- CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/624/workflows/fad2f9bc-b49f-40d6-b3e0-c1c00017607f)
- CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/624/workflows/cd3f93f7-4c9b-4fe1-a993-918fe6b5706c)
CI  lgtm, they are unrelated or known failures.

bereng commented:
@adelapena this [jenkins](https://ci-cassandra.apache.org/job/Cassandra-devbranch/1529/#showFailuresLink) run seems to be good now after the rpm/deb fix. Can you review please?

adelapena commented:
Looks good to me, +1. By the way, regarding [the commit message](https://github.com/apache/cassandra/commit/6946b304e94a8a8d1250680664ddc03b61a139c9), I think it should contain Robert as co-author, since he wrote the original patch.

bereng commented:
^oh wow absolutely my bad. Good catch, Thx.
